docs/errors-and-limits.qmd

```diff
@@ -142,23 +141,16 @@ def intercode_ctf():
 
 This sets a limit of 30 total messages in a conversation before the model is forced to give up. At that point, whatever `output` happens to be in the `TaskState` will be scored (presumably leading to a score of incorrect).
 
-Note that its also possible for a solver to set the `message_limit` directly on the `TaskState` (this is often done by agent solvers which provide their own generate loop):
+Message limits are checked whenever you call `generate()` on any model. The `message_limit` is compared to the number of messages passed in `input` parameter to `generate()`. A `LimitExceededError` will be raised if the message count is equal to or exceeds the limit. This is to avoid proceeding to another (wasteful) generate call if we're already at the limit.
 
-``` python
-@solver
-def agent_loop(message_limit: int = 50):
-    async def solve(state: TaskState, generate: Generate):
+They are also checked on every evaluation of `TaskState.completed`, which will raise a `LimitExceededError` if the message count is equal to or exceeds the limit.
```

Comment:
I am thinking we want to eliminate the `TaskState.completed` check as I think its unnecessary and I don't want to document/emphasize this technique since we want agents to be built w/ `Agent`